// @roots/bud/lib/cjs/extensions/webpack-manifest-plugin/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.when = exports.make = exports.options = exports.name = void 0;
const webpack_manifest_plugin_dependencies_1 = require("./webpack-manifest-plugin.dependencies");
/**
 * Webpack Manifest Plugin adapter
 *
 * @public
 */
const BudWebpackManifestPlugin = {
    /**
     * @public
     */
    name: 'webpack-manifest-plugin',
    /**
     * @public
     */
    options: () => ({
        fileName: 'manifest.json',
        writeToFileEmit: true,
    }),
    /**
     * @public
     */
    make: (options, { hooks }) => {
        return new webpack_manifest_plugin_dependencies_1.WebpackManifestPlugin({
            publicPath: hooks
                .filter('build.output.publicPath')
                .replace('auto', ''),
            ...options.all(),
        });
    },
    /**
     * @public
     */
    when: app => app.store.is('features.manifest', true),
};
exports.name = BudWebpackManifestPlugin.name, exports.options = BudWebpackManifestPlugin.options, exports.make = BudWebpackManifestPlugin.make, exports.when = BudWebpackManifestPlugin.when;
