// @roots/bud-api/lib/cjs/api/methods/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.watch = exports.use = exports.template = exports.extract = exports.splitChunks = exports.serve = exports.runtime = exports.run = exports.proxy = exports.provide = exports.persist = exports.minimize = exports.version = exports.hash = exports.externals = exports.experiments = exports.css = exports.js = exports.entry = exports.devtool = exports.define = exports.override = exports.webpackConfig = exports.config = exports.copy = exports.assets = exports.alias = void 0;
var alias_1 = require("./alias");
Object.defineProperty(exports, "alias", { enumerable: true, get: function () { return alias_1.alias; } });
var assets_1 = require("./assets");
Object.defineProperty(exports, "assets", { enumerable: true, get: function () { return assets_1.method; } });
Object.defineProperty(exports, "copy", { enumerable: true, get: function () { return assets_1.method; } });
var config_1 = require("./config");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return config_1.config; } });
Object.defineProperty(exports, "webpackConfig", { enumerable: true, get: function () { return config_1.config; } });
Object.defineProperty(exports, "override", { enumerable: true, get: function () { return config_1.config; } });
var define_1 = require("./define");
Object.defineProperty(exports, "define", { enumerable: true, get: function () { return define_1.define; } });
var devtool_1 = require("./devtool");
Object.defineProperty(exports, "devtool", { enumerable: true, get: function () { return devtool_1.devtool; } });
var entry_1 = require("./entry");
Object.defineProperty(exports, "entry", { enumerable: true, get: function () { return entry_1.entry; } });
Object.defineProperty(exports, "js", { enumerable: true, get: function () { return entry_1.entry; } });
Object.defineProperty(exports, "css", { enumerable: true, get: function () { return entry_1.entry; } });
var experiments_1 = require("./experiments");
Object.defineProperty(exports, "experiments", { enumerable: true, get: function () { return experiments_1.experiments; } });
var externals_1 = require("./externals");
Object.defineProperty(exports, "externals", { enumerable: true, get: function () { return externals_1.externals; } });
var hash_1 = require("./hash");
Object.defineProperty(exports, "hash", { enumerable: true, get: function () { return hash_1.hash; } });
Object.defineProperty(exports, "version", { enumerable: true, get: function () { return hash_1.hash; } });
var minimize_1 = require("./minimize");
Object.defineProperty(exports, "minimize", { enumerable: true, get: function () { return minimize_1.minimize; } });
var persist_1 = require("./persist");
Object.defineProperty(exports, "persist", { enumerable: true, get: function () { return persist_1.persist; } });
var provide_1 = require("./provide");
Object.defineProperty(exports, "provide", { enumerable: true, get: function () { return provide_1.provide; } });
var proxy_1 = require("./proxy");
Object.defineProperty(exports, "proxy", { enumerable: true, get: function () { return proxy_1.method; } });
var run_1 = require("./run");
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return run_1.run; } });
var runtime_1 = require("./runtime");
Object.defineProperty(exports, "runtime", { enumerable: true, get: function () { return runtime_1.runtime; } });
var serve_1 = require("./serve");
Object.defineProperty(exports, "serve", { enumerable: true, get: function () { return serve_1.method; } });
var splitChunks_1 = require("./splitChunks");
Object.defineProperty(exports, "splitChunks", { enumerable: true, get: function () { return splitChunks_1.method; } });
Object.defineProperty(exports, "extract", { enumerable: true, get: function () { return splitChunks_1.method; } });
var template_1 = require("./template");
Object.defineProperty(exports, "template", { enumerable: true, get: function () { return template_1.template; } });
var use_1 = require("./use");
Object.defineProperty(exports, "use", { enumerable: true, get: function () { return use_1.use; } });
var watch_1 = require("./watch");
Object.defineProperty(exports, "watch", { enumerable: true, get: function () { return watch_1.watch; } });
