// @roots/bud-build/lib/cjs/Build/loaders.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.xml = exports.url = exports.style = exports.resolveUrl = exports.minicss = exports.md = exports.html = exports.file = exports.csv = exports.css = void 0;
const mini_css_extract_plugin_1 = __importDefault(require("mini-css-extract-plugin"));
/**
 * css loader
 */
const css = (app) => app.build.makeLoader().setSrc(require.resolve('css-loader'));
exports.css = css;
/**
 * csv loader
 */
const csv = (app) => app.build.makeLoader().setSrc(require.resolve('csv-loader'));
exports.csv = csv;
const file = (app) => app.build.makeLoader().setSrc(require.resolve('file-loader'));
exports.file = file;
/**
 * Returns {@link Loader} for `html-loader`
 * @public
 */
const html = (app) => app.build.makeLoader().setSrc(require.resolve('html-loader'));
exports.html = html;
/**
 * Returns {@link Loader} for `remark-loader`
 * @public
 */
const md = (app) => app.build.makeLoader().setSrc(require.resolve('remark-loader'));
exports.md = md;
/**
 * Returns {@link Loader} for `mini-css-extract-plugin.loader`
 * @public
 */
const minicss = (app) => app.build.makeLoader().setSrc(mini_css_extract_plugin_1.default.loader);
exports.minicss = minicss;
/**
 * Returns {@link Loader} for `resolve-url-loader`
 * @public
 */
const resolveUrl = (app) => app.build.makeLoader().setSrc(require.resolve('resolve-url-loader'));
exports.resolveUrl = resolveUrl;
/**
 * Returns {@link Loader} for `style-loader`
 * @public
 */
const style = (app) => app.build.makeLoader().setSrc(require.resolve('style-loader'));
exports.style = style;
/**
 * Returns {@link Loader} for `url-loader`
 * @public
 */
const url = (app) => app.build.makeLoader().setSrc(require.resolve('url-loader'));
exports.url = url;
/**
 * Returns {@link Loader} for `xml-loader`
 * @public
 */
const xml = (app) => app.build.makeLoader().setSrc(require.resolve('xml-loader'));
exports.xml = xml;
