// @roots/bud-framework/lib/cjs/Logger/logger.constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.types = exports.INSTANCE_CONFIG = void 0;
const bud_support_1 = require("@roots/bud-support");
/**
 * Instance configuration
 *
 * @internal
 */
exports.INSTANCE_CONFIG = {
    displayScope: true,
    displayBadge: true,
    displayDate: false,
    displayFilename: false,
    displayLabel: false,
    displayTimestamp: false,
    underlineLabel: false,
    underlineMessage: false,
    underlinePrefix: false,
    underlineSuffix: false,
    uppercaseLabel: false,
};
const types = app => ({
    error: {
        badge: bud_support_1.figures.cross,
        color: 'red',
        label: 'error',
        logLevel: "error" /* ERROR */,
    },
    fatal: {
        badge: bud_support_1.figures.cross,
        color: 'red',
        label: 'fatal',
        logLevel: "error" /* ERROR */,
    },
    warn: {
        badge: bud_support_1.figures.warning,
        color: 'yellow',
        label: 'warning',
        logLevel: "error" /* ERROR */,
    },
    star: {
        badge: bud_support_1.figures.star,
        color: 'cyan',
        label: 'instantiate',
        logLevel: "timer" /* STANDARD */,
    },
    info: {
        badge: bud_support_1.figures.info,
        color: 'magenta',
        label: 'log',
        logLevel: "log" /* VERBOSE */,
    },
    success: {
        badge: bud_support_1.figures.tick,
        color: 'green',
        label: 'success',
        logLevel: "timer" /* STANDARD */,
    },
    complete: {
        badge: bud_support_1.figures.circleFilled,
        color: 'cyan',
        label: 'complete',
        logLevel: "timer" /* STANDARD */,
    },
    pending: {
        badge: bud_support_1.figures.ellipsis,
        color: 'cyan',
        label: 'pending',
        logLevel: "timer" /* STANDARD */,
        stream: process.stdout,
    },
    note: {
        badge: bud_support_1.figures.bullet,
        color: 'blue',
        label: 'note',
        logLevel: "timer" /* STANDARD */,
    },
    start: {
        badge: bud_support_1.figures.play,
        color: 'green',
        label: 'start',
        logLevel: "timer" /* STANDARD */,
    },
    pause: {
        badge: bud_support_1.figures.squareSmallFilled,
        color: 'yellow',
        label: 'pause',
        logLevel: "timer" /* STANDARD */,
    },
    debug: {
        badge: bud_support_1.figures.circleFilled,
        color: 'red',
        label: 'log',
        logLevel: "log" /* VERBOSE */,
    },
    await: {
        badge: bud_support_1.figures.ellipsis,
        color: 'cyan',
        label: 'awaiting',
        logLevel: "timer" /* STANDARD */,
    },
    watch: {
        badge: bud_support_1.figures.ellipsis,
        color: 'yellow',
        label: 'watching',
        logLevel: "timer" /* STANDARD */,
    },
    log: {
        badge: bud_support_1.figures.pointer,
        color: 'blue',
        label: 'log',
        logLevel: "timer" /* STANDARD */,
    },
});
exports.types = types;
