// @roots/bud/lib/cjs/factory/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.factory = void 0;
const Bud_1 = require("../Bud");
const context_1 = require("../context");
const seed_1 = require("../seed");
const services_1 = require("../services");
/**
 * Create a {@link Bud} instance programatically
 *
 * @example
 * ```ts
 * const bud = await factory()
 * ```
 *
 * @example
 * Running in a particular mode
 *
 * ```ts
 * const bud = await factory({mode: 'development'})
 * ```
 *
 * @returns Bud instance
 *
 * @public
 */
async function factory(overrides) {
    const context = await (0, context_1.makeContext)();
    const options = {
        name: 'bud',
        mode: 'production',
        ...(overrides ?? {}),
        context: {
            ...context,
            ...(overrides?.context ?? {}),
        },
        services: {
            ...services_1.services,
            ...(overrides?.services ?? {}),
        },
        config: {
            ...seed_1.seed,
            ...(overrides?.config ?? {}),
            location: {
                ...seed_1.seed.location,
                ...(overrides?.config?.location ?? {}),
            },
        },
    };
    const project = new Bud_1.Bud(options);
    process.env.BABEL_ENV = project.mode;
    process.env.NODE_ENV = project.mode;
    project.time(project.name);
    project.log({
        message: 'process.env.NODE_ENV',
        suffix: process.env.NODE_ENV,
    });
    project.log({
        message: 'process.env.BABEL_ENV',
        suffix: process.env.NODE_ENV,
    });
    await project.lifecycle();
    return project;
}
exports.factory = factory;
