// @roots/bud/lib/cjs/context/disk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Disk = void 0;
const bud_support_1 = require("@roots/bud-support");
class Disk {
    constructor(projectDir, config = {}) {
        this.projectDir = projectDir;
        this.config = config;
    }
    async findConfigs() {
        const search = await bud_support_1.globby.globby([
            `*.json`,
            `*.yml`,
            `*.toml`,
            `*.ts`,
            `*.js`,
            `*config*`,
            `*rc`,
            `*lint*`,
            `package.json`,
            `config/*.{ts,js,json,yml}`,
            `!node_modules/**/*`,
        ], {
            absolute: true,
            cwd: this.projectDir,
            dot: true,
            gitignore: true,
            onlyFiles: true,
        });
        this.config = search.reduce((configs, filePath) => ({
            ...configs,
            [`${filePath.split(`${this.projectDir}/`).pop()}`]: filePath,
        }), this.config);
        return this;
    }
}
exports.Disk = Disk;
