// @roots/bud/lib/cjs/index.js
"use strict";
// Copyright © Roots Software Foundation LLC
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.factory = exports.services = exports.extensions = exports.seed = exports.makeContext = exports.Bud = void 0;
/**
 * ⚡️ Frontend build tools combining the best parts of Symfony Encore and Laravel Mix
 *
 * This package provides the Bud class and is the main entrypoint for most projects
 * interfacing with Bud.
 *
 * @see https://roots.io/bud
 * @see https://github.com/roots/bud
 *
 * @packageDocumentation
 */
var Bud_1 = require("./Bud");
Object.defineProperty(exports, "Bud", { enumerable: true, get: function () { return Bud_1.Bud; } });
var context_1 = require("./context");
Object.defineProperty(exports, "makeContext", { enumerable: true, get: function () { return context_1.makeContext; } });
var seed_1 = require("./seed");
Object.defineProperty(exports, "seed", { enumerable: true, get: function () { return seed_1.seed; } });
var extensions_1 = require("./extensions");
Object.defineProperty(exports, "extensions", { enumerable: true, get: function () { return extensions_1.extensions; } });
var services_1 = require("./services");
Object.defineProperty(exports, "services", { enumerable: true, get: function () { return services_1.services; } });
var factory_1 = require("./factory");
Object.defineProperty(exports, "factory", { enumerable: true, get: function () { return factory_1.factory; } });
