// @roots/bud-support/lib/cjs/util/figures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.oneSixth = exports.oneFifth = exports.oneQuarter = exports.oneThird = exports.oneHalf = exports.subscriptNine = exports.subscriptEight = exports.subscriptSeven = exports.subscriptSix = exports.subscriptFive = exports.subscriptFour = exports.subscriptThree = exports.subscriptTwo = exports.subscriptOne = exports.subscriptZero = exports.infinity = exports.identical = exports.greaterOrEqual = exports.lessOrEqual = exports.notEqual = exports.almostEqual = exports.arrowUpDown = exports.arrowLeftRight = exports.arrowRight = exports.arrowLeft = exports.arrowDown = exports.arrowUp = exports.musicNoteBeamed = exports.musicNote = exports.heart = exports.home = exports.triangleRightSmall = exports.triangleLeftSmall = exports.triangleDownSmall = exports.triangleDown = exports.triangleUpSmall = exports.triangleUp = exports.pointerSmall = exports.ellipsis = exports.dot = exports.bullet = exports.squareCenter = exports.squareRight = exports.squareLeft = exports.squareBottom = exports.squareTop = exports.squareLightShade = exports.squareMediumShade = exports.squareDarkShade = exports.square = void 0;
exports.lineDownDoubleLeftDouble = exports.lineDownLeftBold = exports.lineDownBoldLeft = exports.lineDownBoldLeftBold = exports.lineDownLeftArc = exports.lineDownLeft = exports.lineVerticalDashed11 = exports.lineVerticalDashed10 = exports.lineVerticalDashed9 = exports.lineVerticalDashed8 = exports.lineVerticalDashed7 = exports.lineVerticalDashed6 = exports.lineVerticalDashed5 = exports.lineVerticalDashed4 = exports.lineVerticalDashed3 = exports.lineVerticalDashed2 = exports.lineVerticalDashed1 = exports.lineVerticalDashed0 = exports.lineVerticalDouble = exports.lineVerticalBold = exports.lineVertical = exports.lineDashed15 = exports.lineDashed14 = exports.lineDashed13 = exports.lineDashed12 = exports.lineDashed11 = exports.lineDashed10 = exports.lineDashed9 = exports.lineDashed8 = exports.lineDashed7 = exports.lineDashed6 = exports.lineDashed5 = exports.lineDashed4 = exports.lineDashed3 = exports.lineDashed2 = exports.lineDashed1 = exports.lineDashed0 = exports.lineDouble = exports.lineBold = exports.line = exports.sevenEighths = exports.fiveEighths = exports.fiveSixths = exports.fourFifths = exports.threeEighths = exports.threeFifths = exports.threeQuarters = exports.twoFifths = exports.twoThirds = exports.oneEighth = void 0;
exports.lineDownBoldLeftBoldRightBold = exports.lineDownLeftRight = exports.lineUpDownRightDouble = exports.lineUpDoubleDownDoubleRight = exports.lineUpDoubleDownDoubleRightDouble = exports.lineUpBoldDownRight = exports.lineUpDownBoldRight = exports.lineUpDownBoldRightBold = exports.lineUpBoldDownRightBold = exports.lineUpDownRightBold = exports.lineUpBoldDownBoldRight = exports.lineUpBoldDownBoldRightBold = exports.lineUpDownRight = exports.lineUpDownLeftDouble = exports.lineUpDoubleDownDoubleLeft = exports.lineUpDoubleDownDoubleLeftDouble = exports.lineUpBoldDownLeft = exports.lineUpDownBoldLeft = exports.lineUpDownBoldLeftBold = exports.lineUpBoldDownLeftBold = exports.lineUpDownLeftBold = exports.lineUpBoldDownBoldLeft = exports.lineUpBoldDownBoldLeftBold = exports.lineUpDownLeft = exports.lineUpRightDouble = exports.lineUpDoubleRight = exports.lineUpDoubleRightDouble = exports.lineUpRightBold = exports.lineUpBoldRight = exports.lineUpBoldRightBold = exports.lineUpRightArc = exports.lineUpRight = exports.lineUpLeftDouble = exports.lineUpDoubleLeft = exports.lineUpDoubleLeftDouble = exports.lineUpLeftBold = exports.lineUpBoldLeft = exports.lineUpBoldLeftBold = exports.lineUpLeftArc = exports.lineUpLeft = exports.lineDownRightDouble = exports.lineDownDoubleRight = exports.lineDownDoubleRightDouble = exports.lineDownRightBold = exports.lineDownBoldRight = exports.lineDownBoldRightBold = exports.lineDownRightArc = exports.lineDownRight = exports.lineDownLeftDouble = exports.lineDownDoubleLeft = void 0;
exports.circleFilled = exports.circle = exports.squareSmallFilled = exports.squareSmall = exports.cross = exports.warning = exports.info = exports.tick = exports.lineSlash = exports.lineBackslash = exports.lineCross = exports.lineUpDownLeftDoubleRightDouble = exports.lineUpDoubleDownDoubleLeftRight = exports.lineUpDoubleDownDoubleLeftDoubleRightDouble = exports.lineUpDownBoldLeftRightBold = exports.lineUpDownBoldLeftBoldRight = exports.lineUpBoldDownLeftRightBold = exports.lineUpBoldDownLeftBoldRight = exports.lineUpDownLeftBoldRightBold = exports.lineUpBoldDownBoldLeftRight = exports.lineUpDownLeftRightBold = exports.lineUpDownLeftBoldRight = exports.lineUpDownBoldLeftRight = exports.lineUpBoldDownLeftRight = exports.lineUpBoldDownBoldLeftBoldRight = exports.lineUpBoldDownBoldLeftRightBold = exports.lineUpBoldDownLeftBoldRightBold = exports.lineUpDownBoldLeftBoldRightBold = exports.lineUpBoldDownBoldLeftBoldRightBold = exports.lineUpDownLeftRight = exports.lineUpLeftDoubleRightDouble = exports.lineUpDoubleLeftRight = exports.lineUpDoubleLeftDoubleRightDouble = exports.lineUpLeftBoldRight = exports.lineUpLeftRightBold = exports.lineUpBoldLeftRightBold = exports.lineUpBoldLeftBoldRight = exports.lineUpBoldLeftRight = exports.lineUpLeftBoldRightBold = exports.lineUpBoldLeftBoldRightBold = exports.lineUpLeftRight = exports.lineDownLeftDoubleRightDouble = exports.lineDownDoubleLeftRight = exports.lineDownDoubleLeftDoubleRightDouble = exports.lineDownLeftBoldRight = exports.lineDownLeftRightBold = exports.lineDownBoldLeftRightBold = exports.lineDownBoldLeftBoldRight = exports.lineDownBoldLeftRight = exports.lineDownLeftBoldRightBold = void 0;
exports.oneTenth = exports.oneNinth = exports.oneSeventh = exports.nodejs = exports.play = exports.star = exports.mustache = exports.smiley = exports.hamburger = exports.lozengeOutline = exports.lozenge = exports.triangleRight = exports.triangleLeft = exports.triangleUpOutline = exports.pointer = exports.checkboxCircleOff = exports.checkboxCircleOn = exports.checkboxOff = exports.checkboxOn = exports.radioOff = exports.radioOn = exports.circlePipe = exports.circleCross = exports.circleCircle = exports.circleDouble = exports.circleDotted = void 0;
exports.square = '█';
exports.squareDarkShade = '▓';
exports.squareMediumShade = '▒';
exports.squareLightShade = '░';
exports.squareTop = '▀';
exports.squareBottom = '▄';
exports.squareLeft = '▌';
exports.squareRight = '▐';
exports.squareCenter = '■';
exports.bullet = '●';
exports.dot = '․';
exports.ellipsis = '…';
exports.pointerSmall = '›';
exports.triangleUp = '▲';
exports.triangleUpSmall = '▴';
exports.triangleDown = '▼';
exports.triangleDownSmall = '▾';
exports.triangleLeftSmall = '◂';
exports.triangleRightSmall = '▸';
exports.home = '⌂';
exports.heart = '♥';
exports.musicNote = '♪';
exports.musicNoteBeamed = '♫';
exports.arrowUp = '↑';
exports.arrowDown = '↓';
exports.arrowLeft = '←';
exports.arrowRight = '→';
exports.arrowLeftRight = '↔';
exports.arrowUpDown = '↕';
exports.almostEqual = '≈';
exports.notEqual = '≠';
exports.lessOrEqual = '≤';
exports.greaterOrEqual = '≥';
exports.identical = '≡';
exports.infinity = '∞';
exports.subscriptZero = '₀';
exports.subscriptOne = '₁';
exports.subscriptTwo = '₂';
exports.subscriptThree = '₃';
exports.subscriptFour = '₄';
exports.subscriptFive = '₅';
exports.subscriptSix = '₆';
exports.subscriptSeven = '₇';
exports.subscriptEight = '₈';
exports.subscriptNine = '₉';
exports.oneHalf = '½';
exports.oneThird = '⅓';
exports.oneQuarter = '¼';
exports.oneFifth = '⅕';
exports.oneSixth = '⅙';
exports.oneEighth = '⅛';
exports.twoThirds = '⅔';
exports.twoFifths = '⅖';
exports.threeQuarters = '¾';
exports.threeFifths = '⅗';
exports.threeEighths = '⅜';
exports.fourFifths = '⅘';
exports.fiveSixths = '⅚';
exports.fiveEighths = '⅝';
exports.sevenEighths = '⅞';
exports.line = '─';
exports.lineBold = '━';
exports.lineDouble = '═';
exports.lineDashed0 = '┄';
exports.lineDashed1 = '┅';
exports.lineDashed2 = '┈';
exports.lineDashed3 = '┉';
exports.lineDashed4 = '╌';
exports.lineDashed5 = '╍';
exports.lineDashed6 = '╴';
exports.lineDashed7 = '╶';
exports.lineDashed8 = '╸';
exports.lineDashed9 = '╺';
exports.lineDashed10 = '╼';
exports.lineDashed11 = '╾';
exports.lineDashed12 = '−';
exports.lineDashed13 = '–';
exports.lineDashed14 = '‐';
exports.lineDashed15 = '⁃';
exports.lineVertical = '│';
exports.lineVerticalBold = '┃';
exports.lineVerticalDouble = '║';
exports.lineVerticalDashed0 = '┆';
exports.lineVerticalDashed1 = '┇';
exports.lineVerticalDashed2 = '┊';
exports.lineVerticalDashed3 = '┋';
exports.lineVerticalDashed4 = '╎';
exports.lineVerticalDashed5 = '╏';
exports.lineVerticalDashed6 = '╵';
exports.lineVerticalDashed7 = '╷';
exports.lineVerticalDashed8 = '╹';
exports.lineVerticalDashed9 = '╻';
exports.lineVerticalDashed10 = '╽';
exports.lineVerticalDashed11 = '╿';
exports.lineDownLeft = '┐';
exports.lineDownLeftArc = '╮';
exports.lineDownBoldLeftBold = '┓';
exports.lineDownBoldLeft = '┒';
exports.lineDownLeftBold = '┑';
exports.lineDownDoubleLeftDouble = '╗';
exports.lineDownDoubleLeft = '╖';
exports.lineDownLeftDouble = '╕';
exports.lineDownRight = '┌';
exports.lineDownRightArc = '╭';
exports.lineDownBoldRightBold = '┏';
exports.lineDownBoldRight = '┎';
exports.lineDownRightBold = '┍';
exports.lineDownDoubleRightDouble = '╔';
exports.lineDownDoubleRight = '╓';
exports.lineDownRightDouble = '╒';
exports.lineUpLeft = '┘';
exports.lineUpLeftArc = '╯';
exports.lineUpBoldLeftBold = '┛';
exports.lineUpBoldLeft = '┚';
exports.lineUpLeftBold = '┙';
exports.lineUpDoubleLeftDouble = '╝';
exports.lineUpDoubleLeft = '╜';
exports.lineUpLeftDouble = '╛';
exports.lineUpRight = '└';
exports.lineUpRightArc = '╰';
exports.lineUpBoldRightBold = '┗';
exports.lineUpBoldRight = '┖';
exports.lineUpRightBold = '┕';
exports.lineUpDoubleRightDouble = '╚';
exports.lineUpDoubleRight = '╙';
exports.lineUpRightDouble = '╘';
exports.lineUpDownLeft = '┤';
exports.lineUpBoldDownBoldLeftBold = '┫';
exports.lineUpBoldDownBoldLeft = '┨';
exports.lineUpDownLeftBold = '┥';
exports.lineUpBoldDownLeftBold = '┩';
exports.lineUpDownBoldLeftBold = '┪';
exports.lineUpDownBoldLeft = '┧';
exports.lineUpBoldDownLeft = '┦';
exports.lineUpDoubleDownDoubleLeftDouble = '╣';
exports.lineUpDoubleDownDoubleLeft = '╢';
exports.lineUpDownLeftDouble = '╡';
exports.lineUpDownRight = '├';
exports.lineUpBoldDownBoldRightBold = '┣';
exports.lineUpBoldDownBoldRight = '┠';
exports.lineUpDownRightBold = '┝';
exports.lineUpBoldDownRightBold = '┡';
exports.lineUpDownBoldRightBold = '┢';
exports.lineUpDownBoldRight = '┟';
exports.lineUpBoldDownRight = '┞';
exports.lineUpDoubleDownDoubleRightDouble = '╠';
exports.lineUpDoubleDownDoubleRight = '╟';
exports.lineUpDownRightDouble = '╞';
exports.lineDownLeftRight = '┬';
exports.lineDownBoldLeftBoldRightBold = '┳';
exports.lineDownLeftBoldRightBold = '┯';
exports.lineDownBoldLeftRight = '┰';
exports.lineDownBoldLeftBoldRight = '┱';
exports.lineDownBoldLeftRightBold = '┲';
exports.lineDownLeftRightBold = '┮';
exports.lineDownLeftBoldRight = '┭';
exports.lineDownDoubleLeftDoubleRightDouble = '╦';
exports.lineDownDoubleLeftRight = '╥';
exports.lineDownLeftDoubleRightDouble = '╤';
exports.lineUpLeftRight = '┴';
exports.lineUpBoldLeftBoldRightBold = '┻';
exports.lineUpLeftBoldRightBold = '┷';
exports.lineUpBoldLeftRight = '┸';
exports.lineUpBoldLeftBoldRight = '┹';
exports.lineUpBoldLeftRightBold = '┺';
exports.lineUpLeftRightBold = '┶';
exports.lineUpLeftBoldRight = '┵';
exports.lineUpDoubleLeftDoubleRightDouble = '╩';
exports.lineUpDoubleLeftRight = '╨';
exports.lineUpLeftDoubleRightDouble = '╧';
exports.lineUpDownLeftRight = '┼';
exports.lineUpBoldDownBoldLeftBoldRightBold = '╋';
exports.lineUpDownBoldLeftBoldRightBold = '╈';
exports.lineUpBoldDownLeftBoldRightBold = '╇';
exports.lineUpBoldDownBoldLeftRightBold = '╊';
exports.lineUpBoldDownBoldLeftBoldRight = '╉';
exports.lineUpBoldDownLeftRight = '╀';
exports.lineUpDownBoldLeftRight = '╁';
exports.lineUpDownLeftBoldRight = '┽';
exports.lineUpDownLeftRightBold = '┾';
exports.lineUpBoldDownBoldLeftRight = '╂';
exports.lineUpDownLeftBoldRightBold = '┿';
exports.lineUpBoldDownLeftBoldRight = '╃';
exports.lineUpBoldDownLeftRightBold = '╄';
exports.lineUpDownBoldLeftBoldRight = '╅';
exports.lineUpDownBoldLeftRightBold = '╆';
exports.lineUpDoubleDownDoubleLeftDoubleRightDouble = '╬';
exports.lineUpDoubleDownDoubleLeftRight = '╫';
exports.lineUpDownLeftDoubleRightDouble = '╪';
exports.lineCross = '╳';
exports.lineBackslash = '╲';
exports.lineSlash = '╱';
exports.tick = '✔';
exports.info = 'ℹ';
exports.warning = '⚠';
exports.cross = '✖';
exports.squareSmall = '◻';
exports.squareSmallFilled = '◼';
exports.circle = '◯';
exports.circleFilled = '◉';
exports.circleDotted = '◌';
exports.circleDouble = '◎';
exports.circleCircle = 'ⓞ';
exports.circleCross = 'ⓧ';
exports.circlePipe = 'Ⓘ';
exports.radioOn = '◉';
exports.radioOff = '◯';
exports.checkboxOn = '☒';
exports.checkboxOff = '☐';
exports.checkboxCircleOn = 'ⓧ';
exports.checkboxCircleOff = 'Ⓘ';
exports.pointer = '❯';
exports.triangleUpOutline = '△';
exports.triangleLeft = '◀';
exports.triangleRight = '▶';
exports.lozenge = '◆';
exports.lozengeOutline = '◇';
exports.hamburger = '☰';
exports.smiley = '㋡';
exports.mustache = '෴';
exports.star = '★';
exports.play = '▶';
exports.nodejs = '⬢';
exports.oneSeventh = '⅐';
exports.oneNinth = '⅑';
exports.oneTenth = '⅒';
