// @roots/bud-api/lib/cjs/api/methods/template/interpolate-html-plugin.extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BudInterpolateHtmlPlugin = void 0;
const html_webpack_plugin_plugin_1 = require("./html-webpack-plugin.plugin");
const interpolate_html_plugin_plugin_1 = require("./interpolate-html-plugin.plugin");
/**
 * Adapter for {@link InterpolateHtmlPlugin}
 *
 * @public
 */
exports.BudInterpolateHtmlPlugin = {
    /**
     * {@inheritDoc @roots/bud-framework#Extension.CompilerPlugin.name}
     *
     * @public
     */
    name: 'interpolate-html-plugin',
    /**
     * {@inheritDoc @roots/bud-framework#Extension.CompilerPlugin.options}
     *
     * @public
     */
    options: (app) => {
        return {
            ...(app.env.getPublicEnv() ?? {}),
            ...app.extensions.get(`webpack-define-plugin`).options.all(),
        };
    },
    /**
     * {@inheritDoc @roots/bud-framework#Extension.CompilerPlugin.make}
     *
     * @public
     */
    make: (options, app) => {
        app.dump(options.all(), {
            prefix: `${app.name} interpolate-html-plugin made with options`,
        });
        return new interpolate_html_plugin_plugin_1.InterpolateHtmlPlugin(html_webpack_plugin_plugin_1.HtmlWebpackPlugin, options.all(), app);
    },
    /**
     * {@inheritDoc @roots/bud-framework#Extension.CompilerPlugin.when}
     *
     * @public
     */
    when: (_app, options) => options.getEntries().length > 0,
};
